### DISCUSSION TITLE
Tiddly with agent as nullboard

### DISCUSSION THREAD

**Discussion by ghost** [TiddlyWiki/TiddlyWiki5#6858] [Ideas]
#### 1.1 Intro
1. Welcome to TiddlyWiki, a reusable non-linear personal web notebook.  It's a unique [[wiki|WikiWikiWeb]] that people [[love using|Raves]] to keep ideas and information organised. It was originally created by JeremyRuston and is now a thriving [[open source|OpenSourceLicense]] project with a busy [[Community]] of independent developers.
2. TiddlyWiki is written in HTML, CSS and JavaScript to run on any reasonably modern [[browser|Browsers]] without needing any ServerSide logic. It allows anyone to create personal SelfContained hypertext documents that can be published to a web server, sent by email, stored in a Dropbox or kept on a USB thumb drive to make a WikiOnAStick. Because it doesn't need to be installed and configured it makes a great GuerillaWiki.

#### 1.2 feature-name
Tiddly with agent as nullboard

#### 1.3 feature-description 
1. It would be interesting to have an agent, a program that locally saves Tiddly information as the Nullboard - is a program that uses the concept of kanban and is written with html, css, js as Tiddly (In concept). **Note: Nullboard has a similar concept with tiddly with a difference, all data is saved locally by a minimal program called agent, or agent is responsible for saving kanban data.** 
2. Reference here: *The data can also be automatically backed up to a local disk (with the help of [Nullboard Agent](https://nullboard.io/backups)).*

#### 1.4 Problems
1. One of the problems I see with Tiddly is the fact that everything is inside the html/js. If you have a lot of information this can make Tiddly load heavy like it did for me. reference: [1](https://groups.google.com/g/tiddlywiki/c/CfyexUcMeJo), [2](https://groups.google.com/g/Tiddlywiki/c/7ttly1md-OU), [3](https://tiddlywiki.narkive.com/mMRi2oMW/tw-tw5-how-big-can-a-file-be), [4](https://stackoverflow.com/questions/2701039/a-step-up-from-tiddlywiki-that-is-still-100-portable) and [5](https://stackoverflow.com/questions/85994/how-do-you-keep-a-personal-wiki-tiddlywiki-current-and-in-sync-in-multiple-loc)
2. I like to separate things, I prefer to have a separate application layer: frontend, backend.
4. backend on tiddly would be the data layer - the layer responsible for managing, maintaining, versioning my tiddlyers - backend would be the agent I mentioned just now - it would be a program that writes data locally as 'agent' in Nullboard. 

#### 1.5. Why this feature?
1. So...  that's why I mentioned Nullboard because I use it as well as tiddly.
2. Separating application data can improve loading performance.
3. All nullboard data is saved separately which improves performance (loading) and could be implemented with tiddly as well.
4. There are conversations that speak of this perhaps not explicitly - but implicitly, there was a technical discussion about lightweight database like json to manage tiddly internally as well as couchdb and/or pouchdb etc
5. The discussion I'm talking about here would be to separate the tiddly data through an agent program that is responsible for managing, maintaining, updating, creating backups of the tiddly data as Nullboard. *reference here: https://nullboard.io/backups*

#### 2. ux/ui or concept
```html
https://nullboard.io/images/nullboard-backups-4-agent.gif
```

#### 3.1 talks
- https://discourse.joplinapp.org/t/notion-style-and-coda-style-databases/17497
- https://discourse.joplinapp.org/t/has-anyone-migrated-joplin-data-into-the-notion-app/17622
- https://discourse.joplinapp.org/t/notion-to-joplin/19713
- https://discourse.joplinapp.org/t/notion-is-like-an-affair-joplin-is-suitable-for-a-lifetime/5640
- https://talk.tiddlywiki.org/t/notion-like-icon-and-pagecover/1741
- https://talk.tiddlywiki.org/t/multi-user-alternative-to-tiddlywiki/2899
- https://talk.tiddlywiki.org/t/notion-like-backlinks-box/1742/1
- https://groups.google.com/g/tiddlywiki/c/F8QHCRXlqsA/m/9j-Mjyg0CAAJ
- https://talk.tiddlywiki.org/t/multi-user-alternative-to-tiddlywiki/2899
- https://talk.tiddlywiki.org/t/demo-of-a-new-wysiwyg-editor-slate-write-unstable-alpha-stage/2788
- https://talk.tiddlywiki.org/t/how-to-deal-with-wikitext-like-widget-macro-in-wysiwyg-editor/3207
- https://groups.google.com/g/tiddlywiki/c/f23Eof4TTHI/m/RxbryMpGCQAJ
- https://groups.google.com/g/tiddlywiki/c/jKYp7GIm1o8/m/OEE1U_7RAgAJ

#### 3.2 samples
- https://github.com/guiguat/Kanban
- https://github.com/apankrat/nullboard
- https://github.com/Jermolene/TiddlyWiki5
- https://github.com/Jermolene/TiddlyWiki5/discussions/5506
- https://github.com/topics/kanban-board?l=html

#### 3.3 demos
- https://classic.tiddlywiki.com/
- https://wikilabs.github.io/editions/keyvalues/
- https://tiddlywiki.com/dev/
- https://opensource.com/alternatives/trello
- https://leviwheatcroft.github.io/selfhosted-awesome-unlist/taskboard.html
- https://tiddly-gittly.github.io/slate-write/#DemoText:DemoText%20Index%20%24%3A%2Fplugins%2Flinonetwo%2Fslate-write%2Fdevelop%20GettingStarted
- https://gk0wk.github.io/TiddlySeq/#Notion-like%20Icon%20And%20PageCover
- https://tiddly-gittly.github.io/slate-write/#DemoText:DemoText%20Index%20%24%3A%2Fplugins%2Flinonetwo%2Fslate-write%2Fdevelop%20GettingStarted
- https://tiddly-gittly.github.io/slate-write/#DemoText:DemoText%20Index%20%24%3A%2Fplugins%2Flinonetwo%2Fslate-write%2Fdevelop%20GettingStarted
- https://tiddlywiki.com/dev/
- https://nullboard.io/backups
- https://tiddly-gittly.github.io/slate-write/#DemoText:DemoText%20Index%20%24%3A%2Fplugins%2Flinonetwo%2Fslate-write%2Fdevelop%20GettingStarted
- https://wikilabs.github.io/

**Comment by pmario** [comment]
Hello Raphael,
Welcome to our community and to GitHub ;)

Thanks for your feedback and the info about "nullboard". 

TLDR;

All features you mention are already part of TiddlyWiki out of the box. ... The only problem is, that those functions are not obvious. ... So your examples clearly show us, where we have huge potential for improvement in our documentation and presentation of the wiki.

-----------

Wall of text warning ;)

## Some basics about TiddlyWiki

- One of TiddlyWiki's "[unique value propositions](https://en.wikipedia.org/wiki/Unique_selling_proposition)" is, that it is a single-page-application, a [SPA](https://en.wikipedia.org/wiki/Single-page_application), that doesn't need a server to work. 
  - So you can send it by e-mail and it "just works". 
  - Or you can use it with an USB stick ... and so on.
- Since everything is 1 HTML file it will work as long as browsers will exist and they are able to handle Javascript, CSS and HTML code.
  - No other software needed
- Due to it's single file nature [vendor lock in](https://en.wikipedia.org/wiki/Vendor_lock-in) **can** never happen by design.
- [TiddlyWikiClassic](https://classic.tiddlywiki.com/) was created 2004 and it still works 
  - The TiddlyWiki 5 CLI is used to "build" TWclassic ... So the .tid file format has proven to be stable
- In 2014 [Jeremy Ruston wrote](https://tiddlywiki.com/#History%20of%20TiddlyWiki): 

> ... My goal is for it to last for at least 25 years.

heck ... TW5 V1.0 is already 8 years old. ... @Jermolene ... are aware, that TWclassic was born in the [Windows XP era](https://www.thewindowsclub.com/history-of-windows) and has already outlived Windows XP, Windows 7, Windows 8 .. and it will probably still work when Windows 10 is history ;)

I hope that makes it clear that a long-live cycle is a primary goal. 

## from your post

>**1.5 Why this feature?**
>

>2. Separating application data can improve loading performance.

[TiddlyDesktop](https://tiddlywiki.com/#TiddlyDesktop) may be an option here. It can manage single-file wikis and wiki-folders to increase convenience and maintainability.

The TW [CLI](https://en.wikipedia.org/wiki/Command-line_interface) is used to create single-file wikis. That's what you see at tiddlywiki.com ... Where the content and the app is 1 HTML file. We call that [single-file wiki](https://tiddlywiki.com/#SingleFileApplication)

The second possibility is, to start a client-server configuration using the TW CLI and Node.js ... You call this "backend" in your post.

TW uses "folder based" wikis, where every tiddler is 1 `.tid` text file. We call the "data-store" a [wiki folder](https://tiddlywiki.com/#TiddlyWikiFolders), which is mainly used by advanced users and plugin developers. ... 

In our case the main reason to use it that way is maintainability and not speed. 

>3. All nullboard data is saved separately which improves performance (loading) and could be implemented with tiddly as well.

As I wrote. That's already there. But it's not necessarily used to increase performance. We use different saving mechanism for convenience and maintainability. ... The problem depending in several independent elements is increased complexity. ... 

As explained at [nullboard - backups](https://nullboard.io/backups), browsers heavily restrict direct file access from browser based SPAs. 

The app nullboard calls "agent" can roughly be compared with our [Timimi](https://ibnishak.github.io/Timimi/) or [file-backups](https://pmario.github.io/file-backups/) savers. Which allow the wiki to directly save files back to the users file-system. 

Timimi consists of 2 elements. To install the "agent" users need to be allowed to install windows apps, which may be restricted in corporate environments. The second element is a browser-plugin. The advantage of Timimi is, that it can save to every directory the users allows it. --- That's a very similar mechanism as you describe. 

File-backups is a browser plugin only, where the browser limits access to the browser "Downloads" directory and subdirectories. 


>4. There are conversations that speak of this perhaps not explicitly - but implicitly, there was a technical discussion about lightweight database like json to manage tiddly internally as well as couchdb and/or pouchdb etc

There is a TW edition which is named: https://noteself.org which uses couchdb to sync data. 

The discussion about TW data-tiddlers and "database like" behaviour in that context will _not_ improve anything. data-tiddlers are discussed to be deprecated, once we have a better solution. .. A JSON data-store is _no_ database at all. It can be called a  key/value store, but that's already all the commonalities. 

>5. The discussion I'm talking about here would be to separate the tiddly data through an agent program that is responsible for managing, maintaining, updating, creating backups of the tiddly data as Nullboard. reference here: https://nullboard.io/backups

I think that's answered above. 

-----------

## A Rant

TW lacks some convenience features that many users are used today, which can be achieved with client-server configurations. Where the server / backend is controlled by one of the big 5 -- Google, Microsoft, Amazon, Facebook and Apple.

But most of those features have a hefty price tag. Many users trade convenience for privacy and vendor lock-in. 

## Conclusion

TiddlyWiki is free and can be encrypted "out of the box" ... as in free speech ;)

have fun!

**Comment by ghost** [comment]
@pmario Hi! How are you? So... thank you for feedback ;D

I talked about this feature thinking about it here too:

#### 1. concept
![image](https://user-images.githubusercontent.com/110187513/182039706-9d7fd984-ac84-412e-9c02-fba2c12be5e9.png)

#### 2. notes
1. Tiddly looks a lot like the idea: Web Bundle, SingleFile
2. I recently wrote about the importance of a single format for saving sites something similar to the idea of: Web Bundle, SingleFile or Tiddly too - reference: [1](https://discourse.wicg.io/t/integrating-websites-with-the-browser/2512) [2](https://github.com/gildas-lormeau/SingleFile/issues/997) [3](https://github.com/gildas-lormeau/SingleFile/issues/996)
3. I believe projects (Web Bundle, SingleFile or Tiddly): - can solve [this problem :](https://discourse.wicg.io/t/integrating-websites-with-the-browser/2512)

#### 3. question
On a more technical level what is the difference between web-bundles, tiddly and singlepage?

#### 4. references
- https://github.com/gildas-lormeau/SingleFile
- https://github.com/gildas-lormeau/SingleFileZ
- https://web.dev/web-bundles/
- https://www.cdn77.com/blog/web-bundles
- https://wicg.github.io/webpackage/draft-yasskin-wpack-bundled-exchanges.html

**Comment by pmario** [comment]
>On a more technical level what is the difference between web-bundles, tiddly and singlepage?

I don't know web-bundles and singlepage. ... 

Having a very short look at the GitHub repo ...
I think singlepage is a browser addOn, that lets you [scrape](https://en.wikipedia.org/wiki/Web_scraping) a single web-page and it saves it as 1 HTML file. So it seems to load related assets eg: images and the CSS and stores it in a "static webpage". 

TiddlyWiki is fully dynamic and contains many pages, if every tiddler is considered 1 page. TW allows us to edit and then save the content. Depending on the different savers, it's also possible to save changes back to eg: GitHub, GitLab ... and so on. 

Web-bundles, seems to be a new specification coming from Google.. ... I didn't read the spec ... But if it comes from google it will probably be good for Google and users are the product. ... But I may be completely wrong here.

**Comment by amreus** [comment]
`nullboard` is interesting.

`nullboard` is designed to be used from a local file with no server. I do like its saving method, and I think TW might benefit from a similar save strategy. 

`nullboard` uses browser local storage by default. Changes are seamlessly saved in the browser.  Then to persist your board, you select "export" which then saves a json representation of the board as a file in the Downloads folder. If the browser storage is deleted/corrupted, you can import from disk. 

As mentioned, there is also an optional backup server available which automatically saves each revision as you go.
